pkg/sdn/plugin/dns.go

```diff
@@ -0,0 +1,237 @@
+package plugin
+
+import (
+	"fmt"
+	"net"
+	"os/exec"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+
+	kerrors "k8s.io/kubernetes/pkg/util/errors"
+)
+
+const (
+	DIG = "dig"
+
+	defaultTTL = 30 * time.Minute
+)
+
+type dnsValue struct {
+	// All IPv4 addresses for a given domain name
+	ips []net.IP
+	// Time-to-live value from authoritative name server for the domain
+	ttl time.Duration
+	// Holds (last dns lookup time + ttl), tells when to refresh IPs next time
+	nextQueryTime time.Time
+}
+
+type DNS struct {
+	// Protects dnsMap operations
+	lock sync.Mutex
+	// Holds dns name and its corresponding information
+	dnsMap map[string]dnsValue
+}
+
+func CheckDNSResolver() error {
+	if _, err := exec.LookPath(DIG); err != nil {
+		return fmt.Errorf("%s is not installed", DIG)
```

Comment:
If we just log an error instead of returning the error here then this feature will not work and I worry that the cluster admin may not notice this issue until someone tries this feature. I thought, throwing user error here will alert the admin to install the missing library.